cli/internal/fs/copy_file.go

```diff
@@ -28,43 +32,44 @@ func CopyOrLinkFile(from, to string, fromMode, toMode os.FileMode, link, fallbac
 			}
 			return os.Symlink(dest, to)
 		}
-		if err := os.Link(from, to); err == nil || !fallback {
+		if err := os.Link(from.Path.ToString(), to); err == nil || !fallback {
 			return err
 		}
 	}
-	return CopyFile(from, to, toMode)
+	return CopyFile(from, to)
 }
 
 // RecursiveCopy copies either a single file or a directory.
 // 'mode' is the mode of the destination file.
-func RecursiveCopy(from string, to string, mode os.FileMode) error {
-	return RecursiveCopyOrLinkFile(from, to, mode, false, false)
+func RecursiveCopy(from string, to string) error {
+	return RecursiveCopyOrLinkFile(from, to, false, false)
 }
 
 // RecursiveCopyOrLinkFile recursively copies or links a file or directory.
-// 'mode' is the mode of the destination file.
 // If 'link' is true then we'll hardlink files instead of copying them.
 // If 'fallback' is true then we'll fall back to a copy if linking fails.
-func RecursiveCopyOrLinkFile(from string, to string, mode os.FileMode, link, fallback bool) error {
-	info, err := os.Lstat(from)
+func RecursiveCopyOrLinkFile(from string, to string, link bool, fallback bool) error {
+	statedFrom := LstatCachedFile{Path: UnsafeToAbsolutePath(from)}
```

Comment:
Can you add a comment that you verified that the callers do in fact pass in absolute paths?